### PR TITLE
fix regression in router that can cause a crash on response timeout

### LINKED ISSUE #7154: seeing crash starting after syncing with 8d1ad35aa
We synced up with master yesterday (8d1ad35aa) and about an hour ago we saw this crash:

```
Caught Segmentation fault, suspect faulting address 0x0
Backtrace (use tools/stack_decode.py to get line numbers):
#0: __restore_rt [0x7f9b03fde390] ??:0
#1: Envoy::Http::CodecClient::onEvent() [0xa82e4f] /usr/include/c++/7/bits/stl_list.h:962
 (inlined by) external/envoy/source/common/http/codec_client.cc:85
#2: Envoy::Network::ConnectionImpl::raiseEvent() [0x8f8635] /usr/include/c++/7/bits/stl_list.h:162
 (inlined by) external/envoy/source/common/network/connection_impl.cc:332
#3: Envoy::Network::ConnectionImpl::closeSocket() [0x8f7b55] external/envoy/source/common/network/connection_impl.cc:200
#4: Envoy::Network::ConnectionImpl::onReadReady() [0x8f7f7e] external/envoy/source/common/network/connection_impl.cc:521
#5: Envoy::Network::ConnectionImpl::onFileEvent() [0x8f876a] external/envoy/source/common/network/connection_impl.cc:489
#6: Envoy::Event::FileEventImpl::assignEvents()::{lambda()#1}::_FUN() [0x8f239a] external/envoy/source/common/event/file_event_impl.cc:66
#7: event_process_active_single_queue.isra.33 [0xc6ff69] /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/external/com_github_libevent_libevent/event.c:1639
 (inlined by) /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/external/com_github_libevent_libevent/event.c:1698
#8: event_base_loop [0xc71c5f] /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/external/com_github_libevent_libevent/event.c:1806
 (inlined by) /build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/external/com_github_libevent_libevent/event.c:2041
#9: Envoy::Event::DispatcherImpl::run() [0x8f0fca] external/envoy/source/common/event/dispatcher_impl.cc:179
#10: Envoy::Server::WorkerImpl::threadRoutine() [0x8eb32e] bazel-out/k8-opt/bin/external/envoy/source/common/common/_virtual_includes/minimal_logger_lib/common/common/logger.h:276
 (inlined by) external/envoy/source/server/worker_impl.cc:105
#11: Envoy::Thread::ThreadImplPosix::ThreadImplPosix()::{lambda()#1}::_FUN() [0xe74195] external/envoy/source/common/common/posix/thread_impl.cc:41
#12: start_thread [0x7f9b03fd46ba] ??:0
```

I'll dive in to get more info and/or a repro, from a quick look I can't spot any obvious recent PR that would have triggered this.

### PR DESCRIPTION
**Pull request by mpuncel**
Prior to the recent retry hedging change, upstream requests were always
reset as they should be when a response timeout is triggered. The retry
hedging change inadvertently changed the logic so that the upstream
request is only reset if it hadn't seen headers. However it's possible
to hit the timeout after seeing headers and before end_stream which
would result in a crash as the body or trailers are attempted to be
decoded using a null pointer.

Signed-off-by: Michael Puncel <mpuncel@squareup.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Description: Fixes a regression in the router that can cause a crash when an upstream request has seen headers but not end_stream when a response timeout is hit
Risk Level: low
Testing: integration test
Docs Changes: N/A
Release Notes: N/A
Fixes #7154

### CODE REVIEW

File: test/integration/http_timeout_integration_test.cc [hunk 1]
```diff
@@ -47,6 +47,53 @@ TEST_P(HttpTimeoutIntegrationTest, GlobalTimeout) {
...
+  upstream_request_->encodeHeaders(response_headers, false);
+
+  response->waitForHeaders();
+  EXPECT_EQ("200", response->headers().Status()->value().getStringView());
+
+  // Trigger global timeout.
+  timeSystem().sleep(std::chrono::milliseconds(200));
+
+  // Respond with body.
+  upstream_request_->encodeData(100, true);
```
Comment (mpuncel) [inline test/integration/http_timeout_integration_test.cc]: this crashed before the fix was applied

File: test/integration/http_timeout_integration_test.cc [hunk 2]
```diff
@@ -47,6 +47,53 @@ TEST_P(HttpTimeoutIntegrationTest, GlobalTimeout) {
...
+  initialize();
+
+  codec_client_ = makeHttpConnection(makeClientConnection(lookupPort("http")));
+  Http::TestHeaderMapImpl request_headers{{":method", "POST"},
+                                          {":path", "/test/long/url"},
+                                          {":scheme", "http"},
+                                          {":authority", "host"},
+                                          {"x-forwarded-for", "10.0.0.1"},
+                                          {"x-envoy-upstream-rq-timeout-ms", "100"}};
+  auto encoder_decoder = codec_client_->startRequest(request_headers);
```
Comment (lizan) [inline test/integration/http_timeout_integration_test.cc]: use `auto` only if the type is obvious (e.g. cast, make_unique/shared, assigning literals).

**Comment by rgs1** [comment]
Testing this with traffic that triggered the crash for us and thus far, no more crashes 👍

**Comment by lizan** [comment]
The IPv6 test failure seems real, @mpuncel ? 
https://230461-65214191-gh.circle-artifacts.com/0/tmp/envoy-docker/envoy/generated/failed-testlogs/test/integration/http_timeout_integration_test/test.log

**Comment by mpuncel** [comment]
unfortunately it doesn't seem to reproduce in IPv4 tests (even after 600 runs) and I don't have easy access to a machine that can run the IPv6 tests

**Comment by mpuncel** [comment]
Unfortunately I can't seem to reproduce that in gdb or figure out line numbers from stack_decode.py which means I'm kind of stuck

**Comment by mpuncel** [comment]
OK think I lucked out with a guess on what was going on. Now it passes after 1000 runs. Maybe the way the fake upstreams work trying to write a body after a reset is a recipe for badness

**Pull request merged**